### PR TITLE
feat(open): add service URL opener

### LINKED ISSUE #378: Add an open command for service URLs
## Problem
The dashboard shows service links, but terminal workflows still require users to copy ports or URLs from logs before opening an endpoint. This slows down quick checks such as opening the web app, API docs, or a health route.

## Proposed solution
Add ^Gzd app open \<service> [--path \<path>] [--print]. The command resolves the local URL for the service, appends an optional path, and opens it in the default browser. --print writes the URL without opening a browser so scripts can consume it.

## Acceptance criteria
- [ ] ^Gzd app open web opens the detected local URL for the web service.
- [ ] ^Gzd app open api --path /health opens the service URL with the requested path.
- [ ] ^Gzd app open api --print prints the resolved URL and does not launch a browser.
- [ ] Services without a known URL return a clear message with guidance to configure one.
- [ ] Tests cover URL resolution, path joining, print mode, and missing URL cases.

## Notes
Complexity: S · Priority: 3

### PR DESCRIPTION
**Pull request by jongio**
Adds azd app open for opening a resolved local service URL. The command supports --path for routes and --print for scripts, and reports a clear error when a service has no URL source.

Closes #378

Validation: go test ./src/cmd/app/commands -run 'Test(ResolveOpenServiceURLFromCustomURL|ResolveOpenServiceURLFromPorts|ResolveOpenServiceURLMissingURL|JoinOpenURLPath)'

### CODE REVIEW

File: cli/src/cmd/app/commands/open.go [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+		}
+	}
+	return ""
+}
+
+func openURLFromService(svc service.Service) string {
+	if svc.Local != nil && svc.Local.CustomURL != "" {
+		return svc.Local.CustomURL
+	}
+	for _, port := range svc.Ports {
```
Comment (wbreza) [inline cli/src/cmd/app/commands/open.go]: **Medium — Docker container-only ports resolve incorrectly.**

This builds `http://localhost:<port>` from `hostPortFromMapping` without checking whether the service is Docker or whether the extracted value is actually a *published host* port. For a Docker bare port like `"80"`, the container port is exposed on an auto-assigned host port — the canonical `service.ParsePortSpec` returns `HostPort: 0` here, but this code emits `http://localhost:80` (the container port, not reachable on the host).

Prefer `svc.GetPortMappings()` / `svc.GetPrimaryPort()` and only emit a URL when `HostPort > 0`. That also handles IPv6 binds, bind IPs, and protocol suffixes correctly without the manual `strings.Split` logic.

File: cli/src/cmd/app/commands/open.go [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+		if strings.EqualFold(info.Name, serviceName) {
+			foundService = true
+			if base := bestOpenURL(info); base != "" {
+				return joinOpenURLPath(base, extraPath)
+			}
+			break
+		}
+	}
+
+	azureYaml, parseErr := service.ParseAzureYaml(projectDir)
```
Comment (wbreza) [inline cli/src/cmd/app/commands/open.go]: **Low — parse error is swallowed.**

`parseErr` from `ParseAzureYaml` is ignored. Combined with `GetServiceInfo` also suppressing YAML errors, a malformed `azure.yaml` ends up reported as `service %q not found. No services are defined` rather than the real parse failure. Consider surfacing `parseErr` when it is not a simple missing-file case.

File: cli/src/cmd/app/commands/open.go [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+		return base, nil
+	}
+	parsed, err := url.Parse(base)
+	if err != nil {
+		return "", fmt.Errorf("invalid service URL %q: %w", base, err)
+	}
+	if parsed.Scheme == "" || parsed.Host == "" {
+		return "", fmt.Errorf("invalid service URL %q: missing scheme or host", base)
+	}
+	parsed.Path = path.Join(parsed.Path, extraPath)
```
Comment (wbreza) [inline cli/src/cmd/app/commands/open.go]: **Low — `Path` mutated without resetting `RawPath`.**

After `path.Join(parsed.Path, extraPath)`, `parsed.RawPath` still holds the original encoding, so a base with an escaped separator (e.g. `.../a%2Fb`) can render as two segments (`.../a/b/...`) in `parsed.String()`. Also note `path.Join` collapses `..` segments. Reset `parsed.RawPath` to an empty string (or rebuild deliberately) after updating `Path` to keep them consistent.

File: cli/src/cmd/app/commands/open.go [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+	cwd, err := os.Getwd()
+	if err != nil {
+		return fmt.Errorf("failed to get current directory: %w", err)
+	}
+
+	resolved, err := resolveOpenServiceURL(context.Background(), cwd, args[0], openPath)
+	if err != nil {
+		return err
+	}
+	if openPrint {
```
Comment (wbreza) [inline cli/src/cmd/app/commands/open.go:52]: **Nit — output consistency.**

`--print` writes via `fmt.Println`; sibling commands in this package route user-facing output through `internal/output`. Consider using the shared output path for consistency.

**Review by wbreza — COMMENTED** [review]
## Code Review — feat(open): add service URL opener

**Review outcome: Comment.** No Critical/High defects; two Medium and several Low findings. Approving is not warranted, but nothing here is a blocker.

### What this PR does
Adds `azd app open <service>` — resolves a local service URL (running-app state first, then `azure.yaml`) and opens it in the browser, with `--path` to append a route and `--print` to emit the URL instead of launching. Clean, well-tested command with sensible flag seams (`openURL` override for tests).

### Verification note
I built and type-checked the branch against the real packages. All referenced symbols resolve correctly (`serviceinfo.LocalServiceInfo.CustomURL`, `service.Service.Local` -> `LocalServiceConfig.CustomURL`, `completeServiceArgs`, `github.com/pkg/browser`). URL scheme safety is already enforced upstream: `ParseAzureYaml` rejects non-`http(s)` schemes, so `file:`/`javascript:` customUrls from `azure.yaml` cannot reach `browser.OpenURL`. No security finding.

---

### Medium

**1. `go mod tidy` is not committed — preflight/CI will fail.**
`github.com/pkg/browser` is now used directly, but `cli/go.mod` still lists it as `// indirect`. Running `go mod tidy` under `cli/` moves it to the direct `require` block and modifies `go.mod`. A CI "tidy" gate (or the repo's Preflight Checks) will flag this. Run `go mod tidy` in `cli/` and commit the updated `go.mod`/`go.sum`.

**2. Docker container-only ports resolve to the wrong URL.**
`hostPortFromMapping` treats a bare port (e.g. `"80"`) as the host port and emits `http://localhost:80`. For Docker services a bare port is the *container* port with an *auto-assigned* host port — the canonical `service.ParsePortSpec` returns `HostPort: 0` for that case. The hand-rolled parser also skips the Docker/non-Docker distinction and does no numeric validation. Prefer reusing `Service.GetPortMappings()` / `GetPrimaryPort()` and only emit a URL when a real host port is published (`HostPort > 0`). See inline comment.

---

### Low / Nits

- **Swallowed `azure.yaml` parse error** (`resolveOpenServiceURL`): `parseErr` is ignored, so a malformed `azure.yaml` surfaces as "No services are defined" rather than the real error. Consider surfacing non-"missing file" errors. (inline)
- **`joinOpenURLPath` mutates `Path` without resetting `RawPath`**: pre-encoded base paths (e.g. `.../a%2Fb`) can be re-rendered as separate segments (`.../a/b/...`). Queries/fragments are preserved correctly. (inline)
- **Output channel**: `--print` uses `fmt.Println`, while sibling commands route through `internal/output`. Minor consistency nit. (inline)
- **Ignored `context.Context`**: `resolveOpenServiceURL` accepts `ctx` as `_`. Either thread it through the resolvers or drop the parameter.
- **Resolution flow**: when a service is found in service-info but has no URL, the `break` falls through to re-parse `azure.yaml` (which `GetServiceInfo` already parsed). The `foundService` "Start it with 'azd app run'" hint is hard to reach for a service defined in `azure.yaml`. Worth simplifying to a single source of truth.
- **Test gaps**: no coverage for `--print`, the not-found "Available services" list, `hostPortFromMapping` edge cases (3-part `ip:host:container`, `/tcp` suffix, bare port), or path-with-query. Add a table test for `hostPortFromMapping` in particular.

Nice, focused feature — the two Mediums (tidy + docker bare-port) are the ones worth addressing before merge.

File: cli/src/cmd/app/commands/open.go (hunk 2 above)
Comment (jongio) [reply cli/src/cmd/app/commands/open.go]: Agreed — fixed in [`f6faf86`](https://github.com/jongio/azd-app/commit/f6faf86d7a63a1e846e19103598fe8801a85f8f9).

`ParseAzureYaml` has no sentinel error for "file is missing", and I didn't want to classify by substring-matching the message, so I split the two cases explicitly:

```go
func parseOpenAzureYaml(projectDir string) (*service.AzureYaml, error) {
	azureYamlPath, findErr := detector.FindAzureYaml(projectDir)
	if findErr != nil {
		return nil, fmt.Errorf("failed to locate azure.yaml: %w", findErr)
	}
	if azureYamlPath == "" {
		return nil, nil // genuinely absent -> fall through to "service not found"
	}
	azureYaml, parseErr := service.ParseAzureYaml(projectDir)
	if parseErr != nil {
		return nil, fmt.Errorf("failed to parse %s: %w", azureYamlPath, parseErr)
	}
	return azureYaml, nil
}
```

`detector.FindAzureYaml` returns `("", nil)` when it walks to the repo root without finding the file, so a malformed `azure.yaml` now surfaces the real parse failure while a project without one still gets the "No services are defined" message.

One deliberate ordering detail worth calling out: the parse happens **after** the running-app lookup, not before. That way a service that is up and resolvable from live registry state still opens fine even when `azure.yaml` is broken — surfacing the error shouldn't turn a working command into a hard failure.

New test: `TestResolveOpenServiceURLSurfacesParseError` ([`dac97ac`](https://github.com/jongio/azd-app/commit/dac97aca0e55153fc381eb50bdaf41dad2f1b5d8)).


---

_Edited: the branch was rebased onto main to resolve a cli/go.mod conflict with a prometheus/client_golang bump that landed on main. Commit links above point at the post-rebase SHAs. Resolution took main's dependency set plus the direct pkg/browser entry; go mod tidy, go build ./..., the full commands package tests and golangci-lint were all re-run green on the rebased tree._

File: cli/src/cmd/app/commands/open.go (hunk 3 above)
Comment (jongio) [reply cli/src/cmd/app/commands/open.go]: Agreed — fixed in [`f6faf86`](https://github.com/jongio/azd-app/commit/f6faf86d7a63a1e846e19103598fe8801a85f8f9):

```go
	joined := path.Join(parsed.Path, extraPath)
	if strings.HasSuffix(extraPath, "/") && !strings.HasSuffix(joined, "/") {
		joined += "/"
	}
	parsed.Path = joined
	// RawPath is the encoding of the previous Path, so it is stale now. Clear it
	// to keep Path and RawPath consistent and force String() to re-encode.
	parsed.RawPath = ""
```

To be precise about the impact, since I checked before changing it: `URL.EscapedPath()` only returns `RawPath` when it is a valid encoding **of the current `Path`**, so once `Path` changed the stale `RawPath` was already being rejected and the rendered output was usually correct. The real problems were (a) the struct was left internally inconsistent, and (b) there are edge cases — e.g. `--path .`, where `path.Join` cleans back to the original `Path` — in which the stale encoding does win. Clearing it makes `String()` deterministically re-encode.

Added `TestJoinOpenURLPath` cases in [`dac97ac`](https://github.com/jongio/azd-app/commit/dac97aca0e55153fc381eb50bdaf41dad2f1b5d8) for the escaped-separator base (`/a%2Fb`), plus query and fragment preservation, trailing-slash retention, and both error paths.


---

_Edited: the branch was rebased onto main to resolve a cli/go.mod conflict with a prometheus/client_golang bump that landed on main. Commit links above point at the post-rebase SHAs. Resolution took main's dependency set plus the direct pkg/browser entry; go mod tidy, go build ./..., the full commands package tests and golangci-lint were all re-run green on the rebased tree._

File: cli/src/cmd/app/commands/open.go (hunk 4 above)
Comment (jongio) [reply cli/src/cmd/app/commands/open.go:52]: Partially agreed — changed in [`f6faf86`](https://github.com/jongio/azd-app/commit/f6faf86d7a63a1e846e19103598fe8801a85f8f9), but I want to flag a correction on the premise before it becomes received wisdom.

There is no `internal/output` package in this repo — `cli/src/internal/output` doesn't exist — and the sibling commands in this package don't route through a shared output helper either. `env.go`, `health.go`, `outdated.go` and `notifications.go` all call `fmt.Println`/`fmt.Printf` directly, and nothing under `cli/src/cmd/app/commands` uses `cmd.OutOrStdout()`. So `fmt.Println` here was the existing convention, not a deviation from it.

That said, the underlying instinct is right, so I moved `--print` onto the command's own stream:

```go
if _, writeErr := fmt.Fprintln(cmd.OutOrStdout(), resolved); writeErr != nil {
	return fmt.Errorf("failed to write service URL: %w", writeErr)
}
```

Two concrete wins beyond consistency: it makes the `--print` test you asked for in the review body possible (`TestRunOpenPrintWritesURLWithoutBrowser` in [`dac97ac`](https://github.com/jongio/azd-app/commit/dac97aca0e55153fc381eb50bdaf41dad2f1b5d8) captures the buffer and asserts the browser is *not* launched), and it gets the write error checked rather than dropped — `errcheck` flags `fmt.Fprintln` where it exempts `fmt.Println`, which is a fair signal that the silent version was the weaker one.

Leaving this thread open for you: if you'd like the rest of the package migrated to `cmd.OutOrStdout()`, I'm happy to do it, but as its own PR since it touches every command.


---

_Edited: the branch was rebased onto main to resolve a cli/go.mod conflict with a prometheus/client_golang bump that landed on main. Commit links above point at the post-rebase SHAs. Resolution took main's dependency set plus the direct pkg/browser entry; go mod tidy, go build ./..., the full commands package tests and golangci-lint were all re-run green on the rebased tree._

File: cli/src/cmd/app/commands/open.go (hunk 1 above)
Comment (jongio) [reply cli/src/cmd/app/commands/open.go]: Agreed — fixed in [`f6faf86`](https://github.com/jongio/azd-app/commit/f6faf86d7a63a1e846e19103598fe8801a85f8f9).

I dropped the hand-rolled `hostPortFromMapping`/`trimPortProtocol` pair entirely and went through the canonical parser, emitting a URL only when a mapping actually publishes a host port:

```go
func publishedHostPort(svc service.Service) (int, bool) {
	mappings, _ := svc.GetPortMappings()
	for _, mapping := range mappings {
		if mapping.HostPort <= 0 {
			continue
		}
		if mapping.Protocol != "" && !strings.EqualFold(mapping.Protocol, "tcp") {
			continue
		}
		return mapping.HostPort, true
	}
	return 0, false
}
```

Because `GetPortMappings` calls `ParsePortSpec(spec, s.IsContainerService())`, a Docker bare `"80"` now yields `HostPort: 0` and therefore no URL, instead of the unreachable `http://localhost:80`. Bind IPs, IPv6 binds, `/tcp` suffixes and non-numeric specs are all handled the same way the rest of the CLI handles them.

One addition beyond your suggestion: I also skip non-TCP mappings, since a `udp` port isn't reachable over HTTP and the old code would happily have produced `http://localhost:3000` for `"3000:8080/udp"`.

Covered by a table test in `TestOpenURLFromService` ([`dac97ac`](https://github.com/jongio/azd-app/commit/dac97aca0e55153fc381eb50bdaf41dad2f1b5d8)) — docker bare port, docker explicit mapping, `127.0.0.1:3000:8080`, `[::1]:3000:8080`, `/tcp`, `/udp`, malformed spec, and first-publishable-port-wins.


---

_Edited: the branch was rebased onto main to resolve a cli/go.mod conflict with a prometheus/client_golang bump that landed on main. Commit links above point at the post-rebase SHAs. Resolution took main's dependency set plus the direct pkg/browser entry; go mod tidy, go build ./..., the full commands package tests and golangci-lint were all re-run green on the rebased tree._

**Comment by jongio** [comment]
@wbreza thanks — genuinely useful review, and the two Mediums were both real. All four inline threads have replies with commit links; below are the items that only appeared in the review body.

Commits: [`3270f12`](https://github.com/jongio/azd-app/commit/3270f126f28fb804a0b8be112b9d0beb6d4ef1a1) (tidy) · [`f6faf86`](https://github.com/jongio/azd-app/commit/f6faf86d7a63a1e846e19103598fe8801a85f8f9) (fixes) · [`dac97ac`](https://github.com/jongio/azd-app/commit/dac97aca0e55153fc381eb50bdaf41dad2f1b5d8) (tests)

---

> **1. `go mod tidy` is not committed — preflight/CI will fail.**

Confirmed — Preflight was already red on exactly this:

```
Tidying go.mod and go.sum: go.mod or go.sum changed after running go mod tidy
```

Fixed in [`3270f12`](https://github.com/jongio/azd-app/commit/3270f126f28fb804a0b8be112b9d0beb6d4ef1a1): `github.com/pkg/browser` moved from the indirect block into the direct `require` block. `go.sum` was already correct so it is untouched, and re-running `go mod tidy` is now a verified no-op.

---

> **Ignored `context.Context`**: `resolveOpenServiceURL` accepts `ctx` as `_`. Either thread it through or drop the parameter.

Dropped it in [`f6faf86`](https://github.com/jongio/azd-app/commit/f6faf86d7a63a1e846e19103598fe8801a85f8f9). Neither `serviceinfo.GetServiceInfo` nor `service.ParseAzureYaml` takes a `context.Context`, so threading it would have meant changing two package APIs for no present benefit. `runOpen` no longer manufactures a `context.Background()` either.

---

> **Resolution flow**: when a service is found in service-info but has no URL, the `break` falls through to re-parse `azure.yaml` (which `GetServiceInfo` already parsed). ... Worth simplifying to a single source of truth.

Partially agreed, and restructured in [`f6faf86`](https://github.com/jongio/azd-app/commit/f6faf86d7a63a1e846e19103598fe8801a85f8f9) — but the fall-through itself has to stay.

I checked `serviceinfo.mergeServiceInfo`: it copies `local.customUrl` out of `azure.yaml`, but it never derives a URL from `ports:`. So for a service that isn't running, `azure.yaml` is the *only* source that can produce a ports-based URL — collapsing to a single source would break `azd app open api` for any stopped service configured with `ports:` (and `TestResolveOpenServiceURLFromPorts` with it).

What was genuinely broken, and is now fixed:

- the service-info loop no longer `break`s early, so `names` is fully populated before it feeds the "Available services" list — previously a match halfway through truncated it;
- both sources feed one `found` flag and one "no known URL" error, so the `Start it with 'azd app run'` hint is reachable again. You were right that it was effectively dead: the `azure.yaml` branch returned its own error first for any service defined there, which is essentially all of them;
- the `azure.yaml` parse moved *after* the running-state lookup, so a running service still resolves even when `azure.yaml` is malformed.

---

> **Test gaps**: no coverage for `--print`, the not-found "Available services" list, `hostPortFromMapping` edge cases (3-part `ip:host:container`, `/tcp` suffix, bare port), or path-with-query.

All added in [`dac97ac`](https://github.com/jongio/azd-app/commit/dac97aca0e55153fc381eb50bdaf41dad2f1b5d8), and the file is now table-driven testify per `AGENTS.md`:

| Test | Covers |
|---|---|
| `TestRunOpenPrintWritesURLWithoutBrowser` | `--print` writes the URL and does **not** launch a browser |
| `TestRunOpenLaunchesBrowser` | the inverse — browser receives the resolved URL, stdout stays empty |
| `TestResolveOpenServiceURLNotFoundListsServices` | asserts `Available services: api, web` |
| `TestOpenURLFromService` | 11 cases replacing the `hostPortFromMapping` gap: docker bare port, docker explicit, bind IP, IPv6, `/tcp`, `/udp`, malformed spec, multi-port precedence |
| `TestJoinOpenURLPath` | 8 cases: query, fragment, escaped separator, trailing slash, empty path, both error paths |
| `TestResolveOpenServiceURLSurfacesParseError` | malformed `azure.yaml` reports the parse failure |

---

**One correction**, detailed in the `--print` thread: there is no `internal/output` package in this repo and no sibling command uses a shared output helper — `fmt.Println` was the existing convention, not a deviation. I still moved `--print` to `cmd.OutOrStdout()`, because it's the better call and it's what made the `--print` test possible.

**Local verification:** `go build ./...` ✅ · `go test ./src/cmd/app/commands/` (full package) ✅ · `golangci-lint run ./src/cmd/app/commands/...` → 0 issues ✅ · `go mod tidy` → no-op ✅


---

_Edited: the branch was rebased onto main to resolve a cli/go.mod conflict with a prometheus/client_golang bump that landed on main. Commit links above point at the post-rebase SHAs. Resolution took main's dependency set plus the direct pkg/browser entry; go mod tidy, go build ./..., the full commands package tests and golangci-lint were all re-run green on the rebased tree._

**Review by wbreza — APPROVED** [review]
## Code Review (re-review) — feat(open): add service URL opener

**Review outcome: Approve.** All findings from my previous review have been resolved, and the three follow-up commits introduce no new Critical/High/Medium issues. Verified by building and running the tests against the new head.

### Verification
I built the `commands` package and ran the suite in an isolated worktree at `ac7154e`:
- `go build ./src/cmd/app/commands/` — pass
- `go vet ./src/cmd/app/commands/` — pass
- `go test ./src/cmd/app/commands/ -run Open -v` — **18/18 pass**

(The repo-wide dashboard `//go:embed dist` build-order artifact is pre-existing and unrelated to this PR.)

### Prior findings — all addressed
- **Medium — `go mod tidy`**: `pkg/browser` is now a committed direct dependency (`ebacfdfe`). ✅
- **Medium — docker container-only ports**: `publishedHostPort` now defers to the canonical `svc.GetPortMappings()`, skips `HostPort <= 0` (so a docker bare port like `"80"` yields no URL instead of `http://localhost:80`), and filters non-TCP protocols. ✅
- **Low — swallowed `azure.yaml` parse error**: new `parseOpenAzureYaml` uses `detector.FindAzureYaml`, returns `(nil, nil)` for a missing file, and surfaces real parse errors (`TestResolveOpenServiceURLSurfacesParseError`). ✅
- **Low — `RawPath` staleness in `joinOpenURLPath`**: `parsed.RawPath` is now cleared after updating `Path`, with a test asserting re-encoding. ✅
- **Nit — output channel**: `--print` now writes via `fmt.Fprintln(cmd.OutOrStdout(), …)` with write-error handling. ✅
- **Nit — ignored `context`**: the unused `context.Context` parameter was removed. ✅
- **Test gaps**: excellent table-driven coverage added — `TestOpenURLFromService` (11 cases incl. docker bare-port, UDP, IPv6/bind-IP, protocol suffix, malformed spec), `TestJoinOpenURLPath` (query/fragment/re-encode/error cases), `--print` and browser-launch seams, not-found service listing, and parse-error surfacing. ✅

### Optional (non-blocking) nits
- `publishedHostPort` discards the `bool` from `GetPortMappings()` — intentional and correct, since iterating an empty slice is safe.
- `azure.yaml` is parsed twice (once inside `GetServiceInfo`, once in `parseOpenAzureYaml`) — minor redundancy kept for clearer error surfacing; fine as-is.
- Running-state URLs from `bestOpenURL` aren't scheme-validated the way the `azure.yaml` path is (validated upstream by `ParseAzureYaml`); negligible in a local-dev/localhost context.

Clean, well-tested feature that fully incorporates the review feedback. Nice work.

**Review by wbreza — APPROVED** [review]
## Code Review (re-review after rebase) — feat(open): add service URL opener

**Review outcome: Approve.** The only change since my previous approval is a rebase onto the latest `main` (pulling in ~20 merged feature PRs and dependency upgrades). The PR's own contribution is unchanged, and I re-verified it still builds and passes tests against the new base.

### What changed since last review
- Branch rebased onto latest `main` (new head `dac97ac`). The PR's true delta vs the current merge-base is the same four files I already approved: `open.go` (+215), `open_test.go` (+323), `main.go` (+1, registration), `go.mod` (+1/-1, direct `pkg/browser`).
- `open.go` is **byte-for-byte identical** to the version approved at `ac7154e` — no source changes to the feature itself.

### Re-verification against new main (isolated worktree at `dac97ac`)
- `go build ./src/cmd/app/commands/` — pass
- `go build ./src/internal/service/... ./src/internal/serviceinfo/... ./src/internal/detector/...` — pass
- `go vet ./src/cmd/app/commands/` — pass
- `go test ./src/cmd/app/commands/ -run Open -v` — **all Open tests pass (9 functions / 20 subtests)**

The dependency upgrades and `service`/`serviceinfo`/`detector` changes on `main` introduced no API drift: `GetPortMappings() ([]PortMapping, bool)`, `PortMapping.HostPort/Protocol`, `FindAzureYaml`, `LocalServiceInfo.CustomURL/URL/Port`, and `LocalServiceConfig.CustomURL` are all intact and compatible.

(The repo-wide dashboard `//go:embed dist` build-order artifact is pre-existing and unrelated to this PR.)

All findings from earlier rounds remain resolved. Clean rebase — good to merge.

File: cli/src/cmd/app/commands/open.go (hunk 4 above)
Comment (jongio) [reply cli/src/cmd/app/commands/open.go:52]: I dug into this one and I do not think it applies, so I am pushing back rather than changing it.

There is no `internal/output` package in this repo. There is no directory by that name under `cli/src`, no `package output` declaration anywhere, and nothing in `cli/src/cmd/app/commands` imports one.

The `--print` path also is not using `fmt.Println`. `open.go` writes with `fmt.Fprintln(cmd.OutOrStdout(), resolved)` and wraps a write failure in an error, which is already the testable form and already matches how the sibling commands write to the cobra output writer.

Leaving this as is. If you meant a different package, point me at it and I will take another look.
